src/token/TokenUpgrade.sol

```diff
@@ -106,13 +114,18 @@ contract TokenSwap is AccessControl() {
 
     /// @dev Swap a token for its replacement, at the registered ratio. The tokens must have been sent to the contract before this call.
     /// @param tokenIn_ The token to be replaced
-    function swap(IERC20 tokenIn_, address to) external {
+    function swap(IERC20 tokenIn_, address to, bytes32[] calldata proof) external {
         TokenIn memory tokenIn = tokensIn[tokenIn_];
         require(address(tokenIn.reverse) != address(0), "TokenIn not registered");
         IERC20 tokenOut_ = tokenIn.reverse;
 
         uint256 tokenInAmount = tokenIn_.balanceOf(address(this)) - tokenIn.balance;
         uint256 tokenOutAmount = tokenInAmount * tokenIn.ratio / 1e18;
+
+        bytes32 leaf = keccak256(abi.encodePacked(to, tokenOutAmount));
```

Comment:
I'm noticing that there are some conflicts between our pattern for batch operation and this whitelist.

We want to limit holders to upgrade no more than they held at the time of the hack, and that is difficult with the current choice of parameters.

I think we need to do something like:
```
function swap(IERC20 tokenIn_, address from, address to, uint256 tokenInAmount, bytes32[] calldata proof) external {
        TokenIn memory tokenIn = tokensIn[tokenIn_];
        require(address(tokenIn.reverse) != address(0), "TokenIn not registered");
        IERC20 tokenOut_ = tokenIn.reverse;

        tokenIn.safeTransferFrom(from, address(this), tokenInAmount);
        uint256 tokenOutAmount = tokenInAmount * tokenIn.ratio / 1e18;

        bytes32 leaf = keccak256(abi.encodePacked(from, tokenInAmount));
...
```

To allow for batch operation, token holders will have to sign a permit for the TokenUpgrade contract to take their tokens, instead of the regular operation which is to sign the permit for the Ladle.

That also hurts composability a bit because now the TokenUpgrade contract must be the first contract to process tokens in a batch, but fortunately it would be anyway in any case.